lib/spack/spack/schema/concretizer.py

```diff
@@ -9,13 +9,40 @@
 """
 from typing import Any, Dict
 
+LIST_OF_SPECS = {"type": "array", "items": {"type": "string"}}
+
 properties: Dict[str, Any] = {
     "concretizer": {
         "type": "object",
         "additionalProperties": False,
         "properties": {
             "reuse": {
-                "oneOf": [{"type": "boolean"}, {"type": "string", "enum": ["dependencies"]}]
+                "oneOf": [
+                    {"type": "boolean"},
+                    {"type": "string", "enum": ["dependencies"]},
+                    {
+                        "type": "object",
+                        "properties": {
+                            "roots": {"type": "boolean"},
+                            "include": LIST_OF_SPECS,
+                            "exclude": LIST_OF_SPECS,
+                            "from": {
+                                "type": "array",
+                                "items": {
+                                    "type": "object",
+                                    "properties": {
+                                        "type": {
+                                            "type": "string",
+                                            "enum": ["local", "buildcache", "external"],
+                                        },
+                                        "include": LIST_OF_SPECS,
+                                        "exclude": LIST_OF_SPECS,
+                                    },
+                                },
```

Comment:
would really like to be able to write:

```yaml
concretizer:
    reuse:
        from: ["local", "buildcache", "external"]
```

is there an issue w/adding this?

The other thing that is missing from this vs. our idea for `from:` in `packages.yaml` is that there, we were going to specifically allow someone to say `from: "source"`.  How would you say that you *always* want some package to be built from source and never reused here?

Reply:
> How would you say that you always want some package to be built from source and never reused here?

Perhaps we could add a global exclusion?
``` yaml
concretizer:
  reuse:
    roots: true
    from:
    - type: local
      exclude:
      - "openmpi"
    - type: buildcache
      exclude:
      - "openmpi"
    - type: external
      include:
      - "openmpi"
    always_exclude:
    - "end_application_a"
    - "end_application_b"
 ```

Reply:
> How would you say that you always want some package to be built from source and never reused here?

@psakievich @tgamblin There is global exclusion, and imo it's more general than the suggested options:
```yaml
concretizer:
  reuse:
    exclude:
    - "foo"
```
With the above `foo` is not reused from local, buildcache or externals. The only way not to build is to get an accidental hash match. Note that we can also do things like:
```yaml
concretizer:
  reuse:
    exclude:
    - "%oneapi"
```
e.g. rebuild all the specs compiled with `oneapi` etc.

Reply:
🤦‍♂️ well that's what I get for commenting without re-checking the code.

Reply:
ok and `reuse: false` works for everything from source and/or hash?

Reply:
@alalazo: if you do this does it always build from source?  One thing that people had said they wanted was the ability to *always* build a package from source (even if it's installed).  I would not hold this PR up for it but just curious.

`reuse: false` still forces `--fresh` all the time but will do hash reuse, yes?

Reply:
> reuse: false still forces --fresh all the time but will do hash reuse, yes?

Yes. Also, in the case above we'll always do hash reuse (i.e. if the final hash is installed already, we won't override it).

Reply:
@tgamblin would it need to be **always**?  Like if it is an exact hash match why rebuild? The only exception I can think of is for branch based versions, but that should go away when we add the git-sha pairing to branch versions at concretization. I have a minimum viable product working for that in `spack-manager` that I just need to hook up to the concretizer.

Reply:
@psakievich People wanted that, but I don't see why we need to hold up the PR over it.